### PR TITLE
Feature routing work queue tasks

### PR DESCRIPTION
**Pull request by dybi**
Resolves: https://github.com/golemfactory/concent/issues/407

### CODE REVIEW

File: concent_api/core/models.py [hunk 1]
```diff
@@ -306,6 +306,9 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._current_state_name = None
 
+    def __repr__(self):
+        return f"Subtask: task_id={self.task_id}, subtaks_id={self.subtask_id}, state={self.state_enum}"
```
Comment (cameel) [inline concent_api/core/models.py]: `subtaks_id` -> `subtask_id`

File: concent_api/verifier/tasks.py [hunk 2]
```diff
@@ -132,6 +132,9 @@ def verification_result(
             throw=False,
         )
         return
+    except Exception as e:  # pylint: disable=broad-except
+        logging.warning(f"Exception: {e}, Type {type(e)}")
+        return
```
Comment (cameel) [inline concent_api/verifier/tasks.py]: Please remove this overly broad exception handler. I agree with pylint here. If there's a problem, the task should crash. The rule is to crash early and loudly so that we can identify the problems and fix them while still in development and testing. Leave the crash for Celery or Kubernetes to handle. Only catch specific exceptions in specifc situations that you know need to be handled in a special way (like the `DatabaseError` above; though I think that catching all `DatabaseError`s is also a bit too broad there - we only want to handle timeouts not for example errors in the query or in the schema).

You can log it but it's definitely not just a `warning`. But I think the exception will be logged anyway if it escapes from here so it's probably not necessary.

File: concent_api/core/models.py (hunk 1 above)
Comment (cameel) [inline concent_api/core/models.py]: This should be `__str__()`, not `__repr__()`. [repr()](https://docs.python.org/3/library/functions.html#repr) is supposed to return a reprsentation that can be evaluated to get the original object back. `str()` can be more lax.

File: concent_api/concent_api/celery.py [hunk 3]
```diff
@@ -24,7 +24,15 @@
...
     app.conf.imports += tuple(
-        ['{}.tasks'.format(app)
+        [f'{app}.tasks'
          for app in AVAILABLE_CONCENT_FEATURES[feature_name]['required_django_apps']
-         if importlib.util.find_spec('{}.tasks'.format(app)) is not None]
+         if importlib.util.find_spec(f'{app}.tasks') is not None]
     )
+
+app.conf.task_create_missing_queues = True
+app.conf.task_routes = ([
```
Comment (rwrzesien) [inline concent_api/concent_api/celery.py]: What do you think about moving this to settings or constants ?

File: concent_api/concent_api/celery.py [hunk 4]
```diff
@@ -24,7 +24,15 @@
...
+    ('verifier.tasks.verification_result', {'queue': 'concent'}),
+    ('conductor.tasks.blender_verification_request', {'queue': 'conductor'}),
+    ('verifier.tasks.blender_verification_order', {'queue': 'verifier'}),
+],)
```
Comment (rwrzesien) [inline concent_api/concent_api/celery.py]: This has to be tuple ?

File: concent_api/concent_api/celery.py [hunk 5]
```diff
@@ -24,7 +24,15 @@
...
+app.conf.task_default_queue = 'concent'
```
Comment (rwrzesien) [inline concent_api/concent_api/celery.py]: This might also be configurable.

File: concent_api/verifier/tasks.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: I think that the only thing that can go wrong here is `Subtask.DoesNotExists` error.

File: concent_api/concent_api/celery.py (hunk 3 above)
Comment (cameel) [reply concent_api/concent_api/celery.py]: I wanted to suggest this myself but actually I think I like @dybi's solution more. This is something that we never expect the admin to tweak. Keeping it here as a part of code rather than settings makes it pretty clear.

File: concent_api/concent_api/celery.py (hunk 5 above)
Comment (cameel) [reply concent_api/concent_api/celery.py]: It really shouldn't. At least not by someone who's setting up Concent on the cluster.

If there are no bugs in the code this default should never come into play. Each task is explicitly routed to the right queue. If we have a bug and someone forgets to route a task I'd like the code to crash so that we notice it an fix it.

Actually, now that I think about it - maybe we could use a name of a non-existent queue. Would it crash then?

**Comment by cameel** [comment]
One more thing. If #423 goes into master before this pull request please add routing for tasks defined in it.

File: concent_api/verifier/tasks.py (hunk 2 above)
Comment (dybi) [reply concent_api/verifier/tasks.py]: I will remove it.

File: concent_api/core/models.py (hunk 1 above)
Comment (dybi) [reply concent_api/core/models.py]: Not really. 
1) `__repr__` :
> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment).
In this case it is not. 
2) The most important trait of `__repr__` is that it should be unambiguous. Printing information 
about `task_id` and `subtask_id`  is enough to identify a Subtask object.
3) `__str__` of collection (like `QuerySet`) uses `__repr__` of objects it contains (and not their `__str__`).
4) If an object doesn't have `__str__` defined, it will use its `__repr__`, but not vice versa.

**TL;DR**
The use of `__repr__` was intentional.

https://stackoverflow.com/a/2626364

File: concent_api/concent_api/celery.py (hunk 5 above)
Comment (dybi) [reply concent_api/concent_api/celery.py]: This seems a good idea.
@cameel ?

File: concent_api/concent_api/celery.py (hunk 4 above)
Comment (dybi) [reply concent_api/concent_api/celery.py]: yes.
http://docs.celeryproject.org/en/latest/userguide/routing.html#automatic-routing (see: _Note_)

File: concent_api/core/models.py (hunk 1 above)
Comment (dybi) [reply concent_api/core/models.py]: done

File: concent_api/verifier/tasks.py (hunk 2 above)
Comment (dybi) [reply concent_api/verifier/tasks.py]: done

File: concent_api/core/models.py (hunk 1 above)
Comment (cameel) [reply concent_api/core/models.py]: Fair enough. The part about it only having to be unambiguous seems convincing.

But maybe let's tweak the representation a bit so it looks like `Subtask(task_id=1, subtask_id=2, state=3)`? Otherwise it may be hard to read if you print a list of subtasks.

File: concent_api/concent_api/celery.py (hunk 3 above)
Comment (cameel) [reply concent_api/concent_api/celery.py]: As I wrote above, I like this approach and I think we should leave it as is. `celery.py` seems like a very good place for something like this.

File: concent_api/concent_api/celery.py (hunk 5 above)
Comment (cameel) [reply concent_api/concent_api/celery.py]: `task_create_missing_queues` definitely should be disabled. I even stated it in the issue:

> Configure Celery so that new queues are not created automatically.

Think about the use case. When would we ever want a queue to be created automatically? We have a small and constant number of consumer types: concent, conductor and verifier. We know at development time which one should process each task. If won't ever change at runtime. By enabling automatic queue creation the only result is that if you make a typo, your task will go to a new queue that is not being processed by anything. I think it's much, much better if Celery can notify us that we're trying to use a queue that does not exist. If we need a new queue, we'll define it.

File: concent_api/verifier/tasks.py [hunk 6]
```diff
@@ -105,17 +106,21 @@ def blender_verification_order(
...
+    assert isinstance(result, str)
+    assert result in VerificationResult.__members__.keys()
     assert isinstance(error_message, (str, type(None)))
-    assert isinstance(error_code, (ErrorCode, type(None)))
-    assert all([error_message, error_code]) if result == VerificationResult.ERROR else True
+    assert isinstance(error_code, (str, type(None)))
+
+    result_enum = VerificationResult[result]
+
+    assert all([error_message, error_code]) if result_enum == VerificationResult.ERROR else True
```
Comment (cameel) [inline concent_api/verifier/tasks.py]: This could be shortened to:

``` python
assert result_enum != VerificationResult.ERROR or all([error_message, error_code])
```

Comment (cameel) [inline concent_api/verifier/tasks.py]: `[error_message, error_code]` is not a boolean expression. Since it's a non-empty array, it's always true. What is it supposed to check?

File: concent_api/verifier/tasks.py [hunk 7]
```diff
@@ -186,13 +191,13 @@ def verification_result(
...
+    elif result_enum in (VerificationResult.MATCH, VerificationResult.ERROR):
         # Worker logs the error code and message
-        if result == VerificationResult.ERROR:
+        if result_enum == VerificationResult.ERROR:
             logger.info(
                 f'verification_result_task processing error result with: '
-                f'SUBTASK_ID {subtask_id} -- RESULT {result} -- ERROR MESSAGE {error_message} -- ERROR CODE {error_code}'
-            )
+                f'SUBTASK_ID {subtask_id} -- RESULT {result_enum.name} -- ERROR MESSAGE {error_message} -- ERROR CODE {error_code}'
+            )  # pylint: disable=no-member
```
Comment (cameel) [inline concent_api/verifier/tasks.py]: Could you please disable this warning for this enum to `pylintrc`? I see that we're still using `pylint: disable` for enums and it's a really bad practice because it makes us used to disabling warnings rather than treating them as something that has to be fixed.

File: concent_api/verifier/tasks.py (hunk 6 above)
Comment (cameel) [reply concent_api/verifier/tasks.py]: Sorry, I was wrong. It's fine. It does expect a list. I'm too used to the `all(condition for x in collection)` form :)

Comment (rwrzesien) [reply concent_api/verifier/tasks.py]: :+1: done
